sklearn/cluster/tests/test_affinity_propagation.py

```diff
@@ -60,9 +68,31 @@ def test_affinity_propagation():
     assert np.unique(labels).size == n_clusters_
     assert n_clusters == n_clusters_
 
-    # Test also with no copy
+
+def test_affinity_propagation_no_copy():
+    """
+    Check behaviour of not copying the input data.
+    """
+    S = -euclidean_distances(X, squared=True)
+    S_original = S.copy()
+    preference = np.median(S) * 10
+    assert not np.allclose(S.diagonal(), preference)
+
+    # once copy=True S should not be modified
+    affinity_propagation(S, preference=preference, copy=True, random_state=0)
+    assert_allclose(S, S_original)
+
+    # once copy=False S will be modified inplace
+    affinity_propagation(S, preference=preference, copy=False, random_state=0)
+    assert_allclose(S.diagonal(), preference)
```

Comment:
I was expecting this assert to be the inverse of the one on line 83. Is this expectation correct and if yes, does it make sense to re-write the assert like that? (Assuming that future readers will have a similar expectation to me, which is of course speculation :D)

Reply:
> I was expecting this assert to be the inverse of the one on line 83.

Yes, this is also my understanding. 

How would you re-write the assert statement? Can you give a few more details?

Reply:
I was thinking something like `assert not np.allclose(S, S_original)`, to express that `S` should have changed but without making a statement about how it should have changed.

I think internally `assert_allclose()` uses `np.allclose()`, but via a few levels of indirection.